### PR TITLE
DUI: Long names need to be truncated

### PR DESCRIPTION
**Pull request by aosyatnik**
Before:
image

After:
image

Reviewers
@Benglin , please take a look.

Link to YouTrack:
[MAGN-4376](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-4376) DUI: Long names need to be truncated

### CODE REVIEW

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 1]
```diff
@@ -1823,15 +1824,34 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
...
                 case "ClassButton":
-                    text = Dynamo.Nodes.Utilities.InsertSpacesToString(text);
-                    if (text.Length > Configurations.MaxLengthRowClassButtonTitle)
+                    string originalName = Dynamo.Nodes.Utilities.InsertSpacesToString(text);
+                    text = originalName;
+                    int maxRowLength = Configurations.MaxLengthRowClassButtonTitle;
+                    if (text.Length > maxRowLength)
                     {
+                        // If clas name is like "Cooooooool Name", then this truncation is wrong.
                         if (text.IndexOf(" ") != -1)
```
Comment (Benglin) [inline src/DynamoCoreWpf/UI/Converters.cs]: `text.IndexOf(" ")` getting called multiple times when they can be stored in a variable.

File: src/DynamoCoreWpf/UI/Converters.cs [hunk 2]
```diff
@@ -1823,15 +1824,34 @@ public object Convert(object value, Type targetType, object parameter, CultureIn
...
-                            text = text.Insert(text.IndexOf(" ") + 1, "\n");
-                        if (text.Length > Configurations.MaxLengthClassButtonTitle)
+                        {
+                            if (text.IndexOf(" ") <= maxRowLength)
+                                text = text.Insert(text.IndexOf(" ") + 1, "\n");
+                        }
+                        //If class name doesn't have spaces. E.g. "Naaaaaaaaame"
+                        else
+                            text = text.Substring(0, maxRowLength) + Configurations.TwoDots;
```
Comment (Benglin) [inline src/DynamoCoreWpf/UI/Converters.cs]: `Configurations.DoubleDots` may sound better?

**Comment by Benglin** [comment]
Thanks for the fix, Anna. The fix can be further improved on :)

The steps that deal with modifying button text is as follow:

- Insert spaces into original class name:

  - From `SurfaceAnalysisData` --> `Surface Analysis Data`

- Wrap line based on the maximum row size of 8 characters to produce this:

```txt
    Surface
    Analysis
    Data
```

- Merge lines from the back to limit number of rows to `2` and produce this:

```txt
    Surface
    Analysis Data
```

- Append/prefix double dots to the first and last row, depending on the maximum character count:

```txt
    Surface..
    ..Data
```

For each of the steps above, I'd like to see it as individual `Dynamo.Nodes.Utilities` methods that you can make tests for:

```csharp
string InsertSpacesToString(string text);
IEnumerable<string> WrapText(string text, int maxCharacters);
IEnumerable<string> ReduceRowCount(IEnumerable<string> rows, int maxRows);
IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters);
```

Please make sure you add tests that you mentioned in your comments above, also the following:

1. Use `Debug.WriteLine` to output all class names we have in the system, and then pick a few extreme cases as test data (long names, short names, names like "CVS", etc.).
2. Add negative test cases (empty or null names, null arguments, negative counts, long names...).

Does that make sense?

**Comment by aosyatnik** [comment]
Hi, @Benglin , I've created these methods. Please, review it.
Thanks.

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 3]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
             return newText.ToString();
         }
 
+        /// <summary>
+        /// Gets words from text, e.g. ImportFromCSV to ("Import","From","CSV")
+        /// </summary>
+        /// <param name="text">incoming string</param>
+        /// <param name="maxCharacters">Max number of characters per row</param>
+        internal static IEnumerable<string> WrapText(string text, int maxCharacters)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: It seems like `maxCharacters` is not used here, that seems to be a mistake. From the `summary` above, it sounds like this method is the same as `InsertSpacesToString` method.

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 4]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+                return words;
+
+            StringBuilder currentWord = new StringBuilder();
+            currentWord.Append(text[0]);
+            for (int i = 1; i < text.Length; i++)
+            {
+                var curr = text[i];
+                var prev = text[i - 1];
+
+                if (Char.IsUpper(curr) && i != 0 && !Char.IsUpper(prev) && (!curr.Equals(" ")))
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: `i` will never be `0` here.

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 5]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+        /// <summary>
+        /// Gets words from text, e.g. ImportFromCSV to ("Import","From","CSV")
+        /// </summary>
+        /// <param name="text">incoming string</param>
+        /// <param name="maxCharacters">Max number of characters per row</param>
+        internal static IEnumerable<string> WrapText(string text, int maxCharacters)
+        {
+            List<string> words = new List<string>();
+            if (string.IsNullOrWhiteSpace(text))
+                return words;
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: Does `string.IsNullOrWhiteSpace` catch the case where `text` is an empty string? If it doesn't, then the following `text[0]` will crash.

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 6]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+
+            StringBuilder currentWord = new StringBuilder();
+            currentWord.Append(text[0]);
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: The above two lines can be simplified as:

```csharp
var currentWord = new StringBuilder(text[0]);
```

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 7]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+
+        /// <summary>
+        /// Merge rows according max nuber characters per row, e.g. ("Day","Of","Week") => ("Day Of","Week")
+        /// Last row can contain remaining rows, if there is no place to store them,
+        /// e.g. ("Import","From","CSV") => ("Import","From CSV")
+        /// </summary>
+        /// <param name="rows">Incoming rows</param>
+        /// <param name="maxRows">Max number of rows</param>
+        /// <param name="maxCharacters">Max number characters per row</param>
+        internal static IEnumerable<string> ReduceRowCount(IEnumerable<string> rows, int maxRows, int maxCharacters)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: The `maxCharacters` is not required here. The job of this method is to reduce the number of rows, based on the entries inside `rows` parameter. It continuously fold the entry `N - 1` into `N`, until the row count is achieved. For example if the input is `{ "Insert", "Day", "Of", "Week", "Here" }`, and `maxRows == 3` then the iterations would be:

    Iteration 0: { "Insert", "Day", "Of", "Week Here" }
    Iteration 1: { "Insert", "Day", "Of Week Here" }

If `maxRows == 2`, then...

    Iteration 0: { "Insert", "Day", "Of", "Week Here" }
    Iteration 1: { "Insert", "Day", "Of Week Here" }
    Iteration 2: { "Insert", "Day Of Week Here" }

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 8]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+        }
+
+        /// <summary>
+        /// Trincate rows by adding dots, e.g. ("Surface", "Analysis Data") => ("Surface..","..Data")
+        /// </summary>
+        /// <param name="rows">Incomming rows</param>
+        /// <param name="maxCharacters">Max number characters per row</param>
+        internal static IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters)
+        {
+            if (rows == null || maxCharacters <= 0)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: I think `maxCharacter` cannot be less than or equals to `2`, because we need to append two dots, so it must be more than `2` :)

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 9]
```diff
@@ -454,6 +454,147 @@ internal static string InsertSpacesToString(string original)
...
+
+            return resultRows;
+        }
+
+        /// <summary>
+        /// Trincate rows by adding dots, e.g. ("Surface", "Analysis Data") => ("Surface..","..Data")
+        /// </summary>
+        /// <param name="rows">Incomming rows</param>
+        /// <param name="maxCharacters">Max number characters per row</param>
+        internal static IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]:
```csharp
/// <summary>
/// Truncate each entry in the given "rows" to a maximum of "maxCharacters".
/// For examples, given that "maxCharacters" equals to "8":
/// 
///     { "Surface", "Analysis Data" } => { "Surface", "..s Data" }
///     { "By", "Geometry", "Coordinate", "System" } => { "By", "Geometry", "Coordi..", "System" }
///     { "By Geometry", "Coordinate System" } => { "By Geo..", "..System" }
/// 
/// </summary>
/// <param name="rows">...</param>
/// <param name="maxCharacters">...</param>
internal static IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters)
```

**Comment by aosyatnik** [comment]
Hi, @Benglin , I have updated my code. Also I removed `InsertSpacesToString`, since we can use just `WrapText`. Could you, please, take a look?

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 10]
```diff
@@ -430,28 +430,150 @@ internal static string MakeAbsolutePath(string basePath, string relativePath)
...
+        /// Truncate each entry in the given "rows" to a maximum of "maxCharacters".
+        /// For examples, given that "maxCharacters" equals to "8":
+        /// 
+        ///     { "Surface", "Analysis Data" } => { "Surface", "..s Data" }
+        ///     { "By", "Geometry", "Coordinate", "System" } => { "By", "Geometry", "Coordi..", "System" }
+        ///     { "By Geometry", "Coordinate System" } => { "By Geo..", "..System" }
+        /// </summary>
+        /// <param name="rows">Incomming rows</param>
+        /// <param name="maxCharacters">Max number characters per row</param>
+        internal static IEnumerable<string> TruncateRows(IEnumerable<string> rows, int maxCharacters)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: I think the method is overly complicated, can you try something like this (I typed it in Notepad++, so might contains errors):

```csharp
var maxAfterTruncate = maxCharacters - twoDotsLength;
var results = new List<string>();
foreach (var row in rows)
{
    if (row.Length <= maxCharacters) {
        results.Add(row);
        continue;
    }
    
    if (row != lastRow)
    {
        // Rows other than the last get dots appended to the end.
        results.Add(row.SubString(0, maxAfterTruncate) + Configurations.TwoDots);
    }
    else
    {
        // The final row gets the dots added to the front. 
        var offset = row.Length - maxAfterTruncate;
        results.Add(Configurations.TwoDots + row.SubString(offset));
    }
}
```

File: src/DynamoCore/Nodes/NodeCategories.cs [hunk 11]
```diff
@@ -430,28 +430,150 @@ internal static string MakeAbsolutePath(string basePath, string relativePath)
...
+        }
+
+        /// <summary>
+        /// Reduces the number of rows, based on the entries inside rows parameter.
+        /// E.g. rows = { "Insert", "Day", "Of", "Week", "Here" }, maxRows == 3
+        /// Result { "Insert", "Day", "Of Week Here" }
+        /// </summary>
+        /// <param name="rows">Incoming rows</param>
+        /// <param name="maxRows">Max number of rows</param>
+        internal static IEnumerable<string> ReduceRowCount(List<string> rows, int maxRows)
```
Comment (Benglin) [inline src/DynamoCore/Nodes/NodeCategories.cs]: Maybe we won't need to handle the special case of `rows.Count() <= maxRows`, can you try this?

```csharp
var results = new List<string>();
foreach (var row in rows)
{
    // There are still room in the results list.
    if (results.Length < maxRows)
    {
        results.Add(row);
        continue;
    }
    
    // Already full, keep appending to last row.
    var lastRow = results.Last();
    results.RemoveLast();
    results.Add(lastRow + " " + row);
}
```

**Comment by Benglin** [comment]
Thanks @aosyatnik, please update the two methods, and I will check on `WrapText` tomorrow morning, it's getting too late here :)

**Comment by Benglin** [comment]
Thanks @aosyatnik, sorry for taking a month on this one! LGTM.
